### PR TITLE
Bda apply cal

### LINKED ISSUE #510: Enable calibration of BDA data via up/down sampling of calibration solutions in apply_cal
This probably depends on https://github.com/RadioAstronomySoftwareGroup/pyuvdata/issues/428

### PR DESCRIPTION
**Pull request by jsdillon**
This PR enables two different ways of handling BDA data when running `apply_cal`. 

The first is easy: we can choose to upsample or downsample the data to a common temporal resolution (presumably shared with the calibration solution) using the functionality added in #751 to `hera_cal.io`. Applying the calibration solution is then trivial.

This PR also introduces a method for handling BDA by upsampling or downsampling calibration solutions (and associated flags) to match all temporal resolutions in the data. If the calibrations solutions originally have a single time, and are thus trivially broadcastable, we use that. If not, calibrations solutions are iteratively up/downsampled by a factor of 2 in time until all times in the data have corresponding calibration solutions.

This closes #510.

### CODE REVIEW

File: hera_cal/tests/test_apply_cal.py [hunk 1]
```diff
@@ -42,6 +42,77 @@ def test_check_polarization_consistency(self):
         with pytest.raises(KeyError):
             ac._check_polarization_consistency(data, gains)
 
+    def test_build_gains_by_cadences(self):
+        # test upsampling
+        data = {(0, 1, 'nn'): np.ones((8, 10), dtype=complex)}
+        gains = {(0, 'Jnn'): np.array([np.arange(10)]).repeat(2, axis=0).astype(complex)}
+        flags = {(0, 'Jnn'): np.zeros((2, 10), dtype=bool)}
+        gains_by_Nt, cal_flags_by_Nt = ac.build_gains_by_cadences(data, gains, cal_flags=flags)
+        for Nt in [2, 4, 8]:
```
Comment (plaplant) [inline hera_cal/tests/test_apply_cal.py:51]: I think it'd be nice to add testing for non-power-of-two times, once we add the appropriate error/warning handling above.

**Review by plaplant — CHANGES_REQUESTED** [review]
Thanks @jsdillon! This looks great. I just have a few documentation and edge-case handling suggestions. In general, I think we should make assumptions about the shape of the data (e.g., that we only support power-of-two up/downsampling schemes when applying calibrations), and handle other cases more gracefully.

File: hera_cal/tests/test_apply_cal.py (hunk 1 above)
Comment (jsdillon) [reply hera_cal/tests/test_apply_cal.py:51]: I've added warnings and tests for those warnings

File: hera_cal/apply_cal.py [hunk 2]
```diff
@@ -108,12 +108,107 @@ def calibrate_redundant_solution(data, data_flags, new_gains, new_flags, all_red
...
+        gains_by_Nt = {Nt: {} for Nt in data_Nts}
+    else:
+        if np.isscalar(list(gains.values())[0]):
+            gains_by_Nt = {1: {ant: np.array([[gain]]) for ant, gain in gains.items()}}
+        else:
+            gains_by_Nt = {list(gains.values())[0].shape[0]: gains}
+    cal_flags_by_Nt = None
+    if cal_flags is not None:
+        if cal_flags == {}:
+            cal_flags_by_Nt = {Nt: {} for Nt in data_Nts}
```
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/apply_cal.py:151]: Done.

File: hera_cal/apply_cal.py [hunk 3]
```diff
@@ -132,58 +227,79 @@ def calibrate_in_place(data, new_gains, data_flags=None, cal_flags=None, old_gai
...
+
     # loop over baselines in data
     for (i, j, pol) in data.keys():
+
         ap1, ap2 = utils.split_pol(pol)
         flag_all = False
 
+        # get relevant shaped gains for this data waterfall
+        Nt = data[(i, j, pol)].shape[0]
+        new_gains_here = new_gains_by_Nt[Nt]
```
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/apply_cal.py]: Ok, done. And tests have been added.

File: hera_cal/apply_cal.py [hunk 4]
```diff
@@ -132,58 +227,79 @@ def calibrate_in_place(data, new_gains, data_flags=None, cal_flags=None, old_gai
...
+        cal_flags_here = None
+        if cal_flags_by_Nt is not None:
+            cal_flags_here = cal_flags_by_Nt[Nt]
```
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/apply_cal.py]: Done.

File: hera_cal/apply_cal.py [hunk 5]
```diff
@@ -132,58 +227,79 @@ def calibrate_in_place(data, new_gains, data_flags=None, cal_flags=None, old_gai
...
+        old_gains_here = None
+        if old_gains is not None:
+            old_gains_here = old_gains_by_Nt[Nt]
```
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/apply_cal.py]: Done.

File: hera_cal/apply_cal.py [hunk 6]
```diff
@@ -132,58 +227,79 @@ def calibrate_in_place(data, new_gains, data_flags=None, cal_flags=None, old_gai
 
     _check_polarization_consistency(data, new_gains)
     exponent = {'divide': 1, 'multiply': -1}[gain_convention]
+
+    # build dictionary of all necessary gain shapes to account for calibration of BDA data
+    new_gains_by_Nt, cal_flags_by_Nt = build_gains_by_cadences(data, new_gains, cal_flags=cal_flags, flags_are_wgts=flags_are_wgts)
+    if old_gains is not None:
+        old_gains_by_Nt, _ = build_gains_by_cadences(data, old_gains, cal_flags=cal_flags, flags_are_wgts=flags_are_wgts)
```
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/apply_cal.py:246]: I believe this is doable by just replacing `old_gains_by_Nt, _ = build_gains_by_cadences(data, old_gains, cal_flags=cal_flags, flags_are_wgts=flags_are_wgts)` with old_gains_by_Nt, _ = build_gains_by_cadences(data, old_gains)`. So I'll do that.

File: hera_cal/io.py [hunk 7]
```diff
@@ -1574,6 +1574,18 @@ def update_uvdata(uvd, data=None, flags=None, nsamples=None, add_to_history='',
     uvd.check()
 
 
+def _write_HERAData_to_filetype(hd, outfilename, filetype_out='miriad', clobber=False):
+    '''Helper function for update_vis().'''
+    if filetype_out == 'miriad':
+        hd.write_miriad(outfilename, clobber=clobber)
+    elif filetype_out == 'uvfits':
+        hd.write_uvfits(outfilename, force_phase=True, spoof_nonessential=True)
```
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/io.py:1582]: Added a test.

File: hera_cal/apply_cal.py [hunk 8]
```diff
@@ -108,12 +108,107 @@ def calibrate_redundant_solution(data, data_flags, new_gains, new_flags, all_red
                 data[bl] *= avg_gains**exponent
 
 
+def build_gains_by_cadences(data, gains, cal_flags=None, flags_are_wgts=False):
+    ''' Builds dictionaries that map gains to the various cadences in potentially BDA data.
```
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/apply_cal.py:112]: Added a caveat to the docstring.

**Pull request merged**